### DISCUSSION TITLE
How can I alter the default route to go to a specific module in a CMS solution?

### DISCUSSION THREAD

**Discussion by inzi** [OrchardCMS/OrchardCore#10545] [Q&A] [answered]
Scenario - Orchard Core CMS v1.0.0 - 

Nothing should be available to the an unauthenticated user.

I've tried using HomeRoute, attempting to leverage the Account controller's Login method.  The Account controller lives in OrchardCore.Users.

  
       // in recipe

        "HomeRoute": {
        "Action": "Login",
        "Controller": "Account",
        "Area": "OrchardCore.Users",
        "ROUTE_ID": ""
      },

I've tried addControllers() in startup.cs' ConfigureServices method in an attempt to leverage endpoints, which results in MVC throwing that it has multiple matches for "default" (which would be expected as far as I can tell).

So I created a "landing module" and tried:

      "HomeRoute": {
        "Action": "Default",
        "Controller": "Home",
        "Area": "My.Landing",
        "ROUTE_ID": ""
      },

I think this is the better approach, because my landing page could redirect from the controller to login, and upon login, it sends the user back the default homepage.

Any advice is appreciated.

**Comment by deanmarcussen** [comment]
The landing page is a better approach.

If you disable the View All Content permission for the Anonymous role, you will automatically get redirected to the login page if you are not authenticated.

If you are using your own controllers/razor pages, simply apply permissions to the controllers to achieve the same behaviour

**Reply by inzi** [reply]
Thank you very much - this gives me a way to accomplish what I'm trying to do.

My Landing page module, then, will be the "root" of my web application. 

I'll disable the anonymous permissions for content.

I'll replace the login screen in my theme

However - I have been unsuccessful in making the default page (route) of my "landing module" to be the default view using the approaches above.

I think I'm missing a very basic step in accomplishing this. Before trying to secure my landing page, I'm simply trying to make it the default route.  I've found considerable information about this in issues and in documentation, but I've yet to make it work.

The endpoint route ought to be simple, but it throws an error because it can find more than one.   Is the approach to make the landing controller unique within the entire solution? I think that would work, but I'm thinking that's not the best practice. 

In adding the default route, I can specify the controller and the action - but it's not clear how to say (basically) "module=Landing, controller=Home, action=Dashboard"

I hope I'm explaining my question correctly.

**Comment by ns8482e** [comment, marked as answer]
Disable home route feature and provide `/` routing for your controller or page

**Reply by inzi** [reply]
Thank you. I've been investigating accomplishing this based on this discussion, which is exactly what I'm attempting:

https://github.com/OrchardCMS/OrchardCore/issues/6235#issuecomment-648353184

The suggested approach is to use attribute routing or mapareacontrolleroute.

The example given:

> public override void Configure(IApplicationBuilder app, IEndpointRouteBuilder routes, ...)
> {
>     routes.MapAreaControllerRoute(
>         name: "Html.Demo.Test",
>         areaName: "OrchardCore.Html", <= needs to be the module assembly name
>         pattern: "",
>         defaults: new { controller = "Demo", action = "Test" }
>     );
> }

I've tried both approaches, with no success.  I'm liking missing something very basic.

            routes.MapAreaControllerRoute(
                name: "Default",
                areaName: "Flite.Landing",
                pattern: "",
                defaults: new { controller = "Home", action = "Index" }
            );
as well as specifying [Route("")] on the index method of the home controller within my Flite.Landing module.

I have other modules, which also have HomeController(s), but based on my limit knowledge, this should not be an issue?

**Reply by inzi** [reply]
I'm doing this in my home controller within my landing module, but I get a 404.

    [Route("[controller]/[action]")]

    public class HomeController : Controller
    {
        [Route("")]
        [Route("~/")]
        [Route("~/Home")]
        public ActionResult Index()
        {
            return View();
        }
    }

**Reply by inzi** [reply]
That was it. And I had forgotten to add a reference to my module from the solution host.

**Comment by inzi** [comment]
For anyone who runs into this - being new to Orchard Core - my solution was I forgot to add a reference to the module I wanted as default in the host application. 

I was able to make my module be the default by mapping in startup.cs like this:

            routes.MapAreaControllerRoute(
                name: "Home",
                areaName: "Solution.Landing",
                pattern: "",
                defaults: new { controller = "Home", action = "Index" }
            );
